modules/calib3d/src/calibration_handeye.cpp

```diff
@@ -289,32 +289,49 @@ static void calibrateHandEyeTsai(const std::vector<Mat>& Hg, const std::vector<M
     int idx = 0;
     for (size_t i = 0; i < Hg.size(); i++)
     {
-        for (size_t j = i+1; j < Hg.size(); j++, idx++)
+        for (size_t j = i+1; j < Hg.size(); j++)
         {
             //Defines coordinate transformation from Gi to Gj
             //Hgi is from Gi (gripper) to RW (robot base)
             //Hgj is from Gj (gripper) to RW (robot base)
             Mat Hgij = homogeneousInverse(Hg[j]) * Hg[i]; //eq 6
-            vec_Hgij.push_back(Hgij);
             //Rotation axis for Rgij which is the 3D rotation from gripper coordinate frame Gi to Gj
             Mat Pgij = 2*rot2quatMinimal(Hgij);
 
             //Defines coordinate transformation from Ci to Cj
             //Hci is from CW (calibration target) to Ci (camera)
             //Hcj is from CW (calibration target) to Cj (camera)
             Mat Hcij = Hc[j] * homogeneousInverse(Hc[i]); //eq 7
-            vec_Hcij.push_back(Hcij);
             //Rotation axis for Rcij
             Mat Pcij = 2*rot2quatMinimal(Hcij);
 
+            // Discard motions with rotation too small or too close to pi radians
+            double Pgij_norm = cv::norm(Pgij);
+            double Pcij_norm = cv::norm(Pcij);
+            if (Pgij_norm < 0.3 || Pcij_norm < 0.3 || Pgij_norm > 1.7 || Pcij_norm > 1.7) {
```

Comment:
You should add the corresponding angle values in rad or deg in comment, since these values come from the quaternion representation I think?

---

Maybe also add a `CV_LOG_DEBUG` below to inform the user this data has been skipped?

Reply:
Sure, I'll add a comment.

I also feel I should explain how I came to these values.

The 1.7 value, which translates to rotations of about 120 degrees or greater, was based on calculating equation 12 in the paper with the known ground truth `P'_{cg}`. I ran a test with 25 poses, which gives 300 motions (25*24/2). Equation 12 was satisfied in 188 cases. In the other 112 cases, the norm of the `P_{gij}` or `P_{cij}` was 1.7 or greater. The converse was not always true--in the 188 good cases, there were norms greater than 1.7.

The 0.3 value translates to rotations of 17 degrees or less. Honestly, I didn't encounter any problem with these rotations, but I know from the error analysis (and experience) that small rotations are not very informative, so I decided to discard them also.

As for the log statement, I'm willing to add it, but I don't think it's necessary. This implementation tries to use all pairs of poses in the calculation, which is actually including redundant information--once you've used the transform from pose 1 to pose 2, and from pose 2 to pose 3, the transform from pose 1 to pose 3 doesn't add any new information. The log message would be pretty noisy (of course, at DEBUG that's probably OK). I do alert the user if the calibration fails because there's not enough useful data.

Reply:
Thanks for the explanation. Could also be added as comments in the code in my opinion.

OK, no need to add the debug message.